crystal-map-connector/src/main/kotlin/ch/ergon/dope/CrystalMapAdapter.kt

```diff
@@ -1,48 +1,93 @@
+    requireValidConvertable(other.typeConverter.write(this), Number::class).toDopeType()
+
+@JvmName("toDopeTypeListString")
+fun <Convertable : Any> Convertable.toDopeType(other: CMConverterList<Convertable, String>): TypeExpression<StringType> =
+    requireValidConvertable(other.typeConverter.write(this), String::class).toDopeType()
+
+@JvmName("toDopeTypeListBoolean")
+fun <Convertable : Any> Convertable.toDopeType(other: CMConverterList<Convertable, Boolean>): TypeExpression<BooleanType> =
+    requireValidConvertable(other.typeConverter.write(this), Boolean::class).toDopeType()
+
+fun <Convertable : Any, JsonType : Number> CMConverterField<Convertable, JsonType>.toDopeType(other: Convertable) =
+    requireValidConvertable(typeConverter.write(other), Number::class).toDopeType()
+
+fun <Convertable : Any> CMConverterField<Convertable, String>.toDopeType(other: Convertable) =
+    requireValidConvertable(typeConverter.write(other), String::class).toDopeType()
+
+fun <Convertable : Any> CMConverterField<Convertable, Boolean>.toDopeType(other: Convertable) =
+    requireValidConvertable(typeConverter.write(other), Boolean::class).toDopeType()
+
 @JvmName("toDopeNumberField")
-fun CMField<out Number>.toDopeType(reference: String = path): Field<NumberType> = Field(name, reference)
+fun CMJsonField<out Number>.toDopeType(reference: String = path): Field<NumberType> = Field(name, reference)
 
 @JvmName("toDopeStringField")
-fun CMField<String>.toDopeType(reference: String = path): Field<StringType> = Field(name, reference)
+fun CMJsonField<String>.toDopeType(reference: String = path): Field<StringType> = Field(name, reference)
 
 @JvmName("toDopeBooleanField")
-fun CMField<Boolean>.toDopeType(reference: String = path): Field<BooleanType> = Field(name, reference)
+fun CMJsonField<Boolean>.toDopeType(reference: String = path): Field<BooleanType> = Field(name, reference)
 
 @JvmName("toDopeNumberArrayField")
-fun CMList<out Number>.toDopeType(): Field<ArrayType<NumberType>> = Field(name, path)
+fun CMJsonList<out Number>.toDopeType(): Field<ArrayType<NumberType>> = Field(name, path)
 
 @JvmName("toDopeStringArrayField")
-fun CMList<String>.toDopeType(): Field<ArrayType<StringType>> = Field(name, path)
+fun CMJsonList<String>.toDopeType(): Field<ArrayType<StringType>> = Field(name, path)
 
 @JvmName("toDopeBooleanArrayField")
-fun CMList<Boolean>.toDopeType(): Field<ArrayType<BooleanType>> = Field(name, path)
+fun CMJsonList<Boolean>.toDopeType(): Field<ArrayType<BooleanType>> = Field(name, path)
+
+fun CMJsonList<out Any>.toDopeType(): Field<ArrayType<ValidType>> = Field(name, path)
 
 // TODO: DOPE-192
 fun <T : Schema> CMObjectList<T>.toDopeType() = DopeSchemaArray(element, formatPathToQueryString(name, path))
+
+private fun <Convertable : Any, JsonType : Any> Convertable.requireValidConvertable(jsonType: JsonType?, jsonTypeClass: KClass<JsonType>) =
```

Comment:
Is there a test, which tests an invalild (null) conversion?

Reply:
yes it is in the `CrystalMapAdapterTest` and looks like this:
```
@Test
fun `should throw exception when resolving corrupt field`() {
    val string = someString()
    val corruptField = someCorruptField()
    assertThrows<IllegalArgumentException> {
        string.toDopeType(corruptField)
    }
}
```